### PR TITLE
Support for Installing from Edit Page without Commit

### PR DESCRIPTION
**Pull request by Noitidart**
When open a file to edit in GitHub, this will add a button saying checking if its installable. It will look at the file path of the currently open file and then look in its parent directory, then grandparent directory, and so on until it finds an `install.rdf`, when found, it will set the path on the button to this directory. Clicking the install button will download the zip and replace the currently loaded edit page file in the zip with the one from the sourcecode. Allowing to edit on GitHub and commit only when bugs are ironed out. Before this PR, if a user wanted to work out of the GitHub edit file page, they would have to commit every change and then install to test.

`nsIFile.remove`'s were also changed to use `OS.File.remove(nsIFile.path)`

### CODE REVIEW

**Comment by Noitidart** [comment]
hey @diegocr can i get co-author status on AMO so i can upload the next version of this plz, which includes this pull right here, its been some time :P

**Comment by jerone** [comment]
Still useful PR. :shipit:

**Comment by Noitidart** [comment]
Thanks @jerone for this! didn't know others were using it. I rely on this PR!

**Comment by Noitidart** [comment]
Fixed that it wasn't showing button on main page

File: bootstrap.js [hunk 1]
```diff
@@ -267,31 +298,139 @@ function onPageLoad(doc) {
...
 	else if([].some.call(doc.querySelectorAll('table.files > tbody > tr > td.content'),
 		(n) => 'install.rdf' === n.textContent.trim())) {
-
+console.error('here');
 		let c = 7, n, z;
+		n = doc.querySelector('a.sidebar-button[href*=".zip"]');
+		console.error('with just n, n is:', n);
+		addButton(n);
+		/*
 		while(c-- && !(n=doc.querySelector('a.minibutton:nth-child('+c+')')));
```
Comment (jerone) [inline bootstrap.js]: The class `.minibutton` has been replaced by `.btn.btn-sm`.

Comment (Noitidart) [inline bootstrap.js]: Thanks @jerone fixed it all up with your rec, thank you!

File: bootstrap.js [hunk 2]
```diff
@@ -6,6 +6,9 @@
  *
  * Contributor(s):
  *   Diego Casorran <dcasorran@gmail.com> (Original Author)
+ *	 Jerone
+ *	 Zulkarnain K.
+ *	 Noitidart <noitidart@gmail.com> (Authored Contirubtion: "Install from Edit Page without Committing")
```
Comment (jerone) [inline bootstrap.js]: [**@jerone**](https://github.com/jerone) commented on [6 jul. 2014 12:19 CEST](https://github.com/diegocr/GitHubExtIns/pull/10/files#r14574672 "2014-07-06T10:19:38Z - Replied by Github Comment Enhancer"):
> Contirubtion -> Contribution

Still saying `Contirubtion` instead of `Contribution`.

You can also remove my name from the list as I actually didn't committed any code to this addon (see https://github.com/diegocr/GitHubExtIns/pull/8, only the idea was taken, a different approach was used).

**Comment by jerone** [comment]
[**@Noitidart**](https://github.com/Noitidart) commented on [5 apr. 2015 20:13 CEST](https://github.com/diegocr/GitHubExtIns/pull/10#issuecomment-89823584 "2015-04-05T18:13:35Z - Replied by Github Comment Enhancer"):
> Fixed that it wasn't showing button on main page

Working for me :+1:

**Comment by jerone** [comment]
```
<em:version>1.5a4</em:version>
```
Wondering why you are using the Alpha pre-release version identifier on the same version as the released version. This version is lower then the actually released version `1.5` according to [SemVer](http://semver.org/) (see no. 9):
> .9. A pre-release version MAY be denoted by appending a hyphen and a series of dot separated identifiers immediately following the patch version. Identifiers MUST comprise only ASCII alphanumerics and hyphen [0-9A-Za-z-]. Identifiers MUST NOT be empty. Numeric identifiers MUST NOT include leading zeroes. **Pre-release versions have a lower precedence than the associated normal version.** A pre-release version indicates that the version is unstable and might not satisfy the intended compatibility requirements as denoted by its associated normal version. Examples: 1.0.0-alpha, 1.0.0-alpha.1, 1.0.0-0.3.7, 1.0.0-x.7.z.92.

And that's why I getting an update message from the original version.

I would suggest using build metadata (see no. 10):
> .10. Build metadata MAY be denoted by appending a plus sign and a series of dot separated identifiers immediately following the patch or pre-release version. Identifiers MUST comprise only ASCII alphanumerics and hyphen [0-9A-Za-z-]. Identifiers MUST NOT be empty. <ins>Build metadata SHOULD be ignored when determining version precedence</ins>. Thus two versions that differ only in the build metadata, have the same precedence. Examples: 1.0.0-alpha+001, 1.0.0+20130313144700, 1.0.0-beta+exp.sha.5114f85.

**Comment by Noitidart** [comment]
Thanks @jerone  i added you as collaborator please adjust if you have time thanks so much for that. I would do it right away but im in the middle of something right now :(

**Comment by jerone** [comment]
[**@Noitidart**](https://github.com/Noitidart) commented on [6 apr. 2015 19:54 CEST](https://github.com/diegocr/GitHubExtIns/pull/10#issuecomment-90174620 "2015-04-06T17:54:24Z - Replied by Github Comment Enhancer"):
> Thanks [@jerone](https://github.com/jerone)  i added you as collaborator please adjust if you have time thanks so much for that. I would do it right away but im in the middle of something right now :(

Cool. Will have a look.

**Comment by Noitidart** [comment]
Thanks so much!! I love this collaboration with you and diego! :)

**Comment by jerone** [comment]
I love to give the

**Comment by Noitidart** [comment]
Wow wow fantasitc work @jerone thank you!

**Comment by jerone** [comment]
You can revert the last commit about formatting document if you only want bug fixes without formatting changes.

**Comment by Noitidart** [comment]
Thanks @jerone, the beautification is definitely welcome! Thank you!

**Comment by diegocr** [comment]
Hi folks, i'm still alive ;) Thank you both for the great support, even with videos and such, i love that :)

Unfortunately this PR still need to fix some issues - as previously discussed - to get merged, meanwhile @jerone if you would like to create a new PR with ecb1346, 7397ed6 and the button fix i could merge it right away and get you properly credited for that.

Cheers!

**Comment by Noitidart** [comment]
What still needs fixing diego? All the diffs are outdated except the last block of re-used code.

**Comment by jerone** [comment]
@diegocr Great to see you alive again :wink: 

Is it possible for you to review this PR again? @Noitidart (correct me if I'm wrong @Noitidart) and I probably can address your feedback to make this PR merge-ready. I don't have to much time to cherry-pick some commits and make separate PR's for them and then update this PR again. Or you can cherry-pick the commits, I don't care (to much) about credits.

**Comment by Noitidart** [comment]
Absolutely @jerone thanks for asking even though you didn't have to. :)

I feel review/merge of this PR is very needed since last Feb when FlightDeck/AddonBuilder was removed as it provides that same ability of testing till happy then committing. It's been a long time since then so those AddonBuilder users are probably no longer looking for alternatives so three's no more rush but this would be nice to get into the hands of the users.

We'll also have to work out something with code signing addons. as for beta and release we can't install addons like this anymore, maybe we can sign the addons before install.

**Comment by diegocr** [comment]
Okay, nevermind, i manually did it and credited both you :)

As for reviewing this PR again, it is not needed. The discussion is right here above, in a nutshell several duplicated code is in place whose purpose is downloading a whole repository just to check if an install.rdf file exists. Something of this add-on is that it's lightweight, and i'll like to keep it that way, so for this PR to get merged it'll need to get _atomically "re-formatted"_

**Comment by Noitidart** [comment]
Thanks for the credit.

I understand you may not have time to re-review, but all those earlier points I addressed except one. You mentioned "several duplicate code", can you please point those out, as all but one point has been addressed. I'm willing to work with you but I need some communication please, as I'm very confused, I only see one case of duplicated code remaining, all others were resolved. If you see several can you please just make a note as you see it, even a simple "blah" would be great collaboration.

Bottom line I'm doing this to help you and the users. Please help me help you and your users. I get nothing out of  this.

**Comment by Noitidart** [comment]
I get a feeling you don't want this to be apart of the add-on, and it's your add-on, I totally understand if that is the case. I'll just maintain a fork for myself as I find your tool very awesome and very helpful with this feature.

**Comment by diegocr** [comment]
Ok, wrong word again... it might not be _"several"_, don't always take my words too literally :-)

I just mean an important chunk of the functionality code being implemented here, which you also confirmed as not yet addressed in Jul 2014: _"implementd all the discussed stuff. Except that re-use of code in that last block."_

> I get a feeling you don't want this to be apart of the add-on

As i said in my previous post, for this PR to get merged it'll need to get atomically "re-formatted"

I.e: don't do in 40 lines of code what you can do in 10 or less, given it's still properly readable of course.

**Comment by Noitidart** [comment]
Thanks diego for the communication I appreciate it. I only mean the best and am trying to help.

I posted somewhere like in April 2014 (last year) that it was a few lines of code so I didn't address it and was hoping you were ok with this tiny block of repeated code, i cant find the comment now as its been a year. I was hoping for your input on that. I found one comment point out it wasnt addressed: https://github.com/diegocr/GitHubExtIns/pull/10#discussion-diff-12027577

Here is the equivalent in the latest commit by jerone and myself: https://github.com/Noitidart/GitHubExtIns/blob/master/bootstrap.js#L417

I and I'm sure other users of this awesome tool would appreciate if you can help me with that tiny block. It's fewer lines then the earlier cherry picking you had to do yesterday for the class fix I think. I did everything else you asked for, as you can see within that same block I had removed the OS.File and used nsIFile.

**Comment by jerone** [comment]
@Noitidart I think it might be wise to close this PR and try again with the current master. Rebasing this PR probably results in to many merge conflicts.

**Comment by diegocr** [comment]
What you linked on `...#L417` is exactly the bloated block of code i'm referring to.

This is what you're doing there:

 * Checking if the URL contains `/edit/` (we'd also need to verify if this won't cause conflicts in other sections around the site)
 * Check for button existence (duplicated code)
 * Placing a **temporal** button...
 * Some DOM operation to extract paths from the breadcrumb to check install.rdf over :confused:
 * Next you download the whole repository, saving it to the local filesystem as a ZIP file, and then you read/traverse the ZIP **just to check** if it contains an `install.rdf`file... Not only you're using more duplicated code here, what i really dislike is that you're doing all that to check for the existence of a `install.rdf` file in the repository, it's overkill.
 * Finally, if an `install.rdf` file was found in the ZIP, you place the `Install with Uncommitted File` button.

I may end up taking your idea and doing it myself if you don't mind, but certainly my free time is kinda limited these days.

**Comment by Noitidart** [comment]
Thanks @jerone I'll try that. :)

@diegocr thanks for that awesome collab! I really don't mind at all! Long live open source! Thank you!

**Comment by Noitidart** [comment]
Closing in attempt to re-submit PR so diff is easier to see per jerones awesome tip
